src/AdoPat.Test/PatCacheTest.cs

```diff
@@ -0,0 +1,251 @@
+            // Assert
+            // The in-memory cache should contain the PatToken and so should
+            // the underlying storage.
+            cache.GetPat(key).Should().BeEquivalentTo(pat);
+            data.Should().BeEquivalentTo(expectedData);
+        }
+
+        [Test]
+        public void PutCache_ExistingDataInCache()
+        {
+            var key = "needle";
+            var pat = new PatToken
+            {
+                DisplayName = DisplayName,
+                Scope = Scope,
+                TargetAccounts = this.targetAccounts,
+                ValidTo = this.unixEpoch.AddDays(-7),
+                ValidFrom = this.unixEpoch,
+                AuthorizationId = this.authorizationId,
+                Token = Token,
+            };
+
+            // First, write existing data as bytes without our target PatToken.
+            var pats = new Dictionary<string, PatToken> { { "lol", new PatToken { DisplayName = "wut" } } };
+            var existing_data = JsonSerializer.SerializeToUtf8Bytes(pats);
+
+            // Then, add our target PatToken to the expected data for later
+            // comparison.
+            pats.Add(key, pat);
+            var expected_data = JsonSerializer.SerializeToUtf8Bytes(pats);
+
+            var data = new byte[] { };
+            var storage = new Mock<IStorageWrapper>(MockBehavior.Strict);
+            storage.Setup(s => s.ReadData()).Returns(existing_data);
+            storage.Setup(s => s.WriteData(It.IsAny<byte[]>())).Callback((byte[] d) => data = d);
+
+            var cache = new PatCache(storage.Object);
+
+            // Act
+            cache.PutPat(key, pat);
+
+            // Assert
+            // The in-memory cache should contain the PatToken and so should
+            // the underlying storage.
+            cache.GetPat(key).Should().BeEquivalentTo(pat);
+            data.Should().BeEquivalentTo(expected_data);
+        }
+
+        [Test]
+        public void PutCache_OverwriteExistingPatToken()
+        {
+            var key = "needle";
+            var pat = new PatToken
+            {
+                DisplayName = DisplayName,
+                Scope = Scope,
+                TargetAccounts = this.targetAccounts,
+                ValidTo = this.unixEpoch.AddDays(-7),
+                ValidFrom = this.unixEpoch,
+                AuthorizationId = this.authorizationId,
+                Token = Token,
+            };
+
+            // Make our expected PatToken differ only by the Token value.
+            var expectedPat = pat;
+            expectedPat.Token = "New Token Value";
+
+            // First, write existing data as bytes.
+            var pats = new Dictionary<string, PatToken> { { key, pat } };
+            var existing_data = JsonSerializer.SerializeToUtf8Bytes(pats);
+
+            // Then, overwrite the existing data with the expected token.
+            pats[key] = expectedPat;
+            var expected_data = JsonSerializer.SerializeToUtf8Bytes(pats);

```

Comment:
have you named it `expected_data` instead of `expectedData` because of Serialization stuff?

Reply:
Nooope, just forgot which language I was writing. 😆